### PR TITLE
[dev-inspect] Combine to a single end point

### PR DESCRIPTION
**Pull request by tnowacki**
- Merge the transaction mode and move call only mode. Specifying the gas object isn't really necessary for dev inspect

I am not entirely sure how the `read_api` parts of the code should change. Maybe you can help with that @666lcz :) Thanks in advance!

### CODE REVIEW

**Comment by patrickkuo** [comment]
Is this same as `dry_run_transaction` minus the gas object? Do we still need `dry_run_transaction` after this?

File: crates/sui-types/src/gas.rs [hunk 1]
```diff
@@ -221,12 +221,28 @@ impl<'a> SuiGasStatus<'a> {
...
+        computation_gas_unit_price: GasPrice,
+        storage_gas_unit_price: GasPrice,
+    ) -> SuiGasStatus<'a> {
+        let charge = false;
+        Self::new_with_budget_impl(
+            gas_budget,
+            computation_gas_unit_price,
+            storage_gas_unit_price,
+            charge,
         )
```
Comment (amnn) [inline crates/sui-types/src/gas.rs:246]: I'm guessing this pattern is used to label the boolean literal with something more informative, is it common?  On first glance I struggled a bit to catch the difference this function and the one above vs the inline comment style, but I'm happy to get used to this style -- not a big deal.

```suggestion
        Self::new_with_budget_impl(
            gas_budget,
            computation_gas_unit_price,
            storage_gas_unit_price,
            /* charge */ false,
        )
```

**Review by amnn — COMMENTED** [review]
Looks good from the authority/adapter side.  Holding off on the accept to give @666lcz a chance to look at the API side.  I'm guessing we'll need to make changes to the SDK(s) downstream of the RPC changes as well?  Will leave it between you to navigate that.

File: crates/sui-types/src/gas.rs (hunk 1 above)
Comment (tnowacki) [reply crates/sui-types/src/gas.rs:246]: I've switched back and forth with both, so I'm not sure what is preferred. I sometimes like the variable name one only in that I find it easier to move around (and it sometimes is better for formatting) though I don't think that either case comes into play here

**Comment by tnowacki** [comment]
> Is this same as `dry_run_transaction` minus the gas object? Do we still need `dry_run_transaction` after this? @gegaowp ?

It is not. Dry run runs object ownership checks, entry rule checks, valid usage of `Pure`, etc. Dry run should be (long term) equivalent to running the exact transaction possibly with/without gas.  

Dev inspect is a literally any transaction with any arguments. 

Mentioned in the meeting this morning, but copying here for posterity. We can split that up in a few different ways, but from a client side, it might make sense to expose dev-inspect and dry-run with different flags

File: crates/sui-types/src/messages.rs [hunk 2]
```diff
@@ -406,6 +406,8 @@ impl SingleTransactionKind {
         // transaction.
         let mut used = HashSet::new();
         if !input_objects.iter().all(|o| used.insert(o.object_id())) {
+            dbg!(input_objects);
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply crates/sui-types/src/messages.rs:409]: Oops :)

File: crates/sui-core/src/unit_tests/authority_tests.rs [hunk 3]
```diff
@@ -275,9 +274,10 @@ async fn test_dev_inspect_object_by_bytes() {
     .await
     .unwrap();
     assert_eq!(effects.created.len(), 1);
-    assert_eq!(effects.mutated.len(), 1);
+    // gas is not mutated
+    assert!(effects.mutated.is_empty());
     assert!(effects.deleted.is_empty());
-    assert_eq!(effects.mutated[0].reference.object_id, gas_object_id);
+    assert!(effects.gas_used.computation_cost > 0);
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply crates/sui-core/src/unit_tests/authority_tests.rs:280]: I think you should get the same result? Let me test it :)

File: crates/sui-core/src/authority.rs [hunk 4]
```diff
@@ -1108,20 +1085,21 @@ impl AuthorityState {
             .storage_gas_price
             .into();
         let gas_status =
-            SuiGasStatus::new_with_budget(MAX_TX_GAS, storage_gas_price, storage_gas_price);
-        let (effects, execution_result) =
-            execution_engine::manual_execute_move_call::<execution_mode::DevInspect, _>(
+            SuiGasStatus::new_uncharged(MAX_TX_GAS, GasPrice::from(1), storage_gas_price);
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply crates/sui-core/src/authority.rs:1088]: On which part, the storage gas? or the computation price of 1? 

We could take these things in as parameters from the user if that would be better?

File: crates/sui-core/src/unit_tests/authority_tests.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tnowacki) [reply crates/sui-core/src/unit_tests/authority_tests.rs:280]: Testing this btw opened up a bunch of issues now fixed. Sorry the PR is a bit drastically different after this

File: crates/sui-core/src/authority.rs [hunk 5]
```diff
@@ -1085,26 +1085,44 @@ impl AuthorityState {
         SuiTransactionEffects::try_from(effects, self.module_cache.as_ref())
     }
 
+    /// The object ID for gas can be any object ID, even for an uncreated object
```
Comment (amnn) [inline crates/sui-core/src/authority.rs:1088]: Does this comment still make sense? This function doesn't accept a gas object ID.

File: crates/sui-core/src/authority.rs [hunk 6]
```diff
@@ -1076,59 +1085,45 @@ impl AuthorityState {
...
             &self.database,
-            input_objects,
-        )?;
+            &transaction_kind,
+            gas_object,
+        )
+        .await?;
         let shared_object_refs = input_objects.filter_shared_objects();
 
+        // TODO should we error instead for 0?
```
Comment (amnn) [inline crates/sui-core/src/authority.rs:1114]: I'm a fan of the existing solution (picking a sensible default).

File: crates/sui-core/src/unit_tests/authority_tests.rs [hunk 7]
```diff
@@ -320,13 +320,14 @@ async fn test_dev_inspect_object_by_bytes() {
         .unwrap()
         .contents()
         .to_vec();
+    // gas used should be the same
+    let actual_gas_used: SuiGasCostSummary = effects.gas_used.into();
+    assert_eq!(actual_gas_used, dev_inspect_gas_summary);
```
Comment (amnn) [inline crates/sui-core/src/unit_tests/authority_tests.rs:325]: 🎉

**Review by amnn — APPROVED** [review]
This looks good to me!  Will accept to unblock, but cc @666lcz in case there are changes necessary on the SDK side

File: crates/sui-core/src/authority.rs (hunk 5 above)
Comment (tnowacki) [reply crates/sui-core/src/authority.rs:1088]: Nope, sorry, was from an intermediate version of this PR

File: crates/sui-json-rpc/src/read_api.rs [hunk 8]
```diff
@@ -239,48 +230,21 @@ impl SuiRpcModule for ReadApi {
...
-
-    async fn dev_inspect_move_call(
         &self,
         sender_address: SuiAddress,
-        package_object_id: ObjectID,
-        module: String,
-        function: String,
-        type_arguments: Vec<SuiTypeTag>,
-        arguments: Vec<SuiJsonValue>,
+        tx_bytes: Base64,
```
Comment (666lcz) [inline crates/sui-json-rpc/src/read_api.rs:235]: hmm I need to take a closer look at this - the problem is that on the client side there's no easy way to serialize data into `TransactionKind` - we only expose API for constructing `TransactionData`

**Review by gegaowp — APPROVED** [review]
Great move, the read_api part looks good per se, will need proper BCS encoding indeed for txn kind. Maybe we can further merge dry run and dev inspect to one.

File: crates/sui-json-rpc/src/read_api.rs [hunk 9]
```diff
@@ -239,48 +230,21 @@ impl SuiRpcModule for ReadApi {
...
+        gas_price: Option<u64>,
```
Comment (666lcz) [inline crates/sui-json-rpc/src/read_api.rs:236]: In api.rs, `gas_price` is after `epoch` but here it is in front of `epoch`. I'll fix this in my diff

File: crates/sui-json-rpc/src/read_api.rs [hunk 10]
```diff
@@ -239,48 +230,21 @@ impl SuiRpcModule for ReadApi {
...
-                type_arguments,
-                arguments,
-            )
-            .await?;
+        let tx_kind: TransactionKind =
+            bcs::from_bytes(&tx_bytes.to_vec().map_err(|e| anyhow!(e))?).map_err(|e| anyhow!(e))?;
         Ok(self
             .state
-            .dev_inspect_move_call(sender_address, move_call, epoch)
+            .dev_inspect_transaction(sender_address, tx_kind, gas_price.unwrap_or(1), epoch)
```
Comment (666lcz) [inline crates/sui-json-rpc/src/read_api.rs:247]:
```suggestion
            .dev_inspect_transaction(sender_address, tx_kind, gas_price.unwrap_or(1), epoch)
```

we should use reference gas price here as the default. I'll make this change in my diff as well unless there's any concern

**Review by 666lcz — CHANGES_REQUESTED** [review]
@tnowacki I have added TS Support. However, one of the e2e test is failing with the following error:

```
test/e2e/dev-inspect.test.ts > Test dev inspect > Dev inspect transaction with PayAllSui
Error: Error dev inspect transaction with request type: Error: RPC Error: SUI payment transactions use first input coin for gas payment, but found a different gas object.
 ❯ JsonRpcProvider.devInspectTransaction src/providers/json-rpc-provider.ts:737:13
    735|       return resp;
    736|     } catch (err) {
    737|       throw new Error(
       |             ^
    738|         `Error dev inspect transaction with requ…
    739|       );
```

Due to this check https://github.com/MystenLabs/sui/blob/cc0c5bf0326a04e9613d5f8eea5e508ccecf2d64/crates/sui-types/src/messages.rs#L973 . Should we bypass this for DevInspect?

File: sdk/typescript/src/providers/json-rpc-provider.ts [hunk 11]
```diff
@@ -706,11 +699,36 @@ export class JsonRpcProvider extends Provider {
...
+    sender: SuiAddress,
+    tx: UnserializedSignableTransaction | string | Base64DataBuffer,
+    gasPrice: number | null = null,
+    epoch: number | null = null
+  ): Promise<DevInspectResults> {
     try {
+      const version = await this.getRpcApiVersion();
+      // TODO: remove after 0.23.0 is deployed in both DevNet and TestNet
+      if (version?.major == 0 && version?.minor < 23) {
+        return this.devInspectTransactionDeprecated(sender, tx, epoch);
```
Comment (666lcz) [inline sdk/typescript/src/providers/json-rpc-provider.ts:712]: @tnowacki We are going to release a new DevNet on Jan 25th(Wednesday). If this is not merged before the DevNet release, then we would need to update `version?.minor < 23` to `version?.minor < 24`.

Comment (tnowacki) [reply sdk/typescript/src/providers/json-rpc-provider.ts]: When is the cut for 0.23?

**Review by 666lcz — APPROVED** [review]

**Pull request merged**